neqo-transport/src/connection/mod.rs

```diff
@@ -3047,23 +3067,34 @@ impl Connection {
         )
     }
 
+    #[allow(clippy::too_many_arguments)]
     fn handle_ack<R>(
         &mut self,
         space: PacketNumberSpace,
+        next_pn: PacketNumber,
         largest_acknowledged: PacketNumber,
         ack_ranges: R,
         ack_ecn: Option<EcnCount>,
         ack_delay: u64,
         now: Instant,
-    ) where
+    ) -> Res<()>
+    where
         R: IntoIterator<Item = RangeInclusive<PacketNumber>> + Debug,
         R::IntoIter: ExactSizeIterator,
     {
         qdebug!([self], "Rx ACK space={}, ranges={:?}", space, ack_ranges);
-
         let Some(path) = self.paths.primary() else {
-            return;
+            return Ok(());
         };
+
+        // Ensure that the largest acknowledged packet number was actually sent.
+        // (If we ever start using non-contigous packet numbers, we need to check all the packet
+        // numbers in the ACKed ranges.)
+        if largest_acknowledged >= next_pn {
+            qwarn!("Largest ACKed {} was never sent", largest_acknowledged,);
+            return Err(Error::AckedUnsentPacket);
+        }
```

Comment:
Instead of validating the untrusted value `largest_acknowledged`, can we instead derive it ourself, looking at the largest packet in `acked_packets` below, after calling `self.loss_recovery.on_ack_received`?

`acked_packets` should only contain packets we have sent before and thus a `largest_acknowledged` derived from `acked_packets` doesn't need the validation above.

Reply:
https://github.com/larseggert/neqo/pull/31 implements this proposal @larseggert. Note the PR description:

> Note that this pull request does not consider the remote behavior as malicious. It simply ignores the largest ACK, if it doesn't correspond to a packet previously sent.